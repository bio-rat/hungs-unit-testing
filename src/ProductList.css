.product-item {
  background-color: black;
  color: white;
  /* width: 200px; */
  margin-top: 10px;
}

body {
  margin: 0 auto;
}
