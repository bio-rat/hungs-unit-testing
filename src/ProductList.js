import React from "react";
import "./ProductList.css";

const ProductList = ({ products, onProductBuy }) => {
  return (
    <div>
      {products.map(x => (
        <li className="product-item" key={x.id}>
          {x.name} : ${x.price} USD
          <button onClick={() => onProductBuy(x.id)}>Add to Card</button>
          <h1>Hello</h1>
        </li>
      ))}
    </div>
  );
};

export default ProductList;
