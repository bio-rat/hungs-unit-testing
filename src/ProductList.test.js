import React from "react";
import { shallow } from "enzyme";
import ProductList from "./ProductList";

let TEST_DATA;
let output;
let testingFn = jest.fn();

beforeEach(() => {
  TEST_DATA = [
    { id: 134, name: "computer", price: 1000 },
    { id: 122, name: "anker mouse", price: 50.25 },
    { id: 553, name: "computer stand", price: 20.99 },
    { id: 689, name: "mac adapter", price: 10000 }
  ];
  output = shallow(
    <ProductList onProductBuy={testingFn} products={TEST_DATA} />
  );
});

it("Render a list of products", () => {
  expect(output.find("li").length).toEqual(4);
});

it("Render price as $XXX.XX USD but not show .00", () => {
  const output = shallow(<ProductList products={TEST_DATA} />);
  expect(output.text()).toMatch(/\$\d{1,} USD/ || !/[.]00/);
});

it("Each li has to have className", () => {
  const list = output.find("li");
  list.forEach(node => {
    expect(node.hasClass("product-item")).toEqual(true);
  });
});

it("Each li has a button", () => {
  const firstProduct = output.find("li").first();
  expect(firstProduct.find("button").length).toBe(1);
});

it("Each button has text 'Add to Card' ", () => {
  expect(
    output
      .find("button")
      .first()
      .contains("Add to Card")
  ).toBe(true);
});

it("Click button works", () => {
  const firstProduct = output.find("button").first();
  firstProduct.simulate("click");
  expect(testingFn.mock.calls.length).toEqual(1);
});

it("Check params through function", () => {
  const firstProduct = output.find("button").first();
  firstProduct.simulate("click");
  expect(testingFn.mock.calls[0][0]).toEqual(134);
});

it("Check snapshot", () => {
  expect(output).toMatchSnapshot();
});
