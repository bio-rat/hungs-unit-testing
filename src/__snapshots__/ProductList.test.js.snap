// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProductList
    onProductBuy={
      [MockFunction] {
        "calls": Array [
          Array [
            134,
          ],
          Array [
            134,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    products={
      Array [
        Object {
          "id": 134,
          "name": "computer",
          "price": 1000,
        },
        Object {
          "id": 122,
          "name": "anker mouse",
          "price": 50.25,
        },
        Object {
          "id": 553,
          "name": "computer stand",
          "price": 20.99,
        },
        Object {
          "id": 689,
          "name": "mac adapter",
          "price": 10000,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          className="product-item"
        >
          computer
           : $
          1000
           USD
          <button
            onClick={[Function]}
          >
            Add to Card
          </button>
          <h1>
            Hello
          </h1>
        </li>,
        <li
          className="product-item"
        >
          anker mouse
           : $
          50.25
           USD
          <button
            onClick={[Function]}
          >
            Add to Card
          </button>
          <h1>
            Hello
          </h1>
        </li>,
        <li
          className="product-item"
        >
          computer stand
           : $
          20.99
           USD
          <button
            onClick={[Function]}
          >
            Add to Card
          </button>
          <h1>
            Hello
          </h1>
        </li>,
        <li
          className="product-item"
        >
          mac adapter
           : $
          10000
           USD
          <button
            onClick={[Function]}
          >
            Add to Card
          </button>
          <h1>
            Hello
          </h1>
        </li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "134",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "computer",
            " : $",
            1000,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>,
            <h1>
              Hello
            </h1>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "computer",
          " : $",
          1000,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to Card",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to Card",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello",
            },
            "ref": null,
            "rendered": "Hello",
            "type": "h1",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "122",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "anker mouse",
            " : $",
            50.25,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>,
            <h1>
              Hello
            </h1>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "anker mouse",
          " : $",
          50.25,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to Card",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to Card",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello",
            },
            "ref": null,
            "rendered": "Hello",
            "type": "h1",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "553",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "computer stand",
            " : $",
            20.99,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>,
            <h1>
              Hello
            </h1>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "computer stand",
          " : $",
          20.99,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to Card",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to Card",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello",
            },
            "ref": null,
            "rendered": "Hello",
            "type": "h1",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "689",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "mac adapter",
            " : $",
            10000,
            " USD",
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>,
            <h1>
              Hello
            </h1>,
          ],
          "className": "product-item",
        },
        "ref": null,
        "rendered": Array [
          "mac adapter",
          " : $",
          10000,
          " USD",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add to Card",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add to Card",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Hello",
            },
            "ref": null,
            "rendered": "Hello",
            "type": "h1",
          },
        ],
        "type": "li",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            className="product-item"
          >
            computer
             : $
            1000
             USD
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>
            <h1>
              Hello
            </h1>
          </li>,
          <li
            className="product-item"
          >
            anker mouse
             : $
            50.25
             USD
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>
            <h1>
              Hello
            </h1>
          </li>,
          <li
            className="product-item"
          >
            computer stand
             : $
            20.99
             USD
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>
            <h1>
              Hello
            </h1>
          </li>,
          <li
            className="product-item"
          >
            mac adapter
             : $
            10000
             USD
            <button
              onClick={[Function]}
            >
              Add to Card
            </button>
            <h1>
              Hello
            </h1>
          </li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "134",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "computer",
              " : $",
              1000,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to Card
              </button>,
              <h1>
                Hello
              </h1>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "computer",
            " : $",
            1000,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to Card",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to Card",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Hello",
              },
              "ref": null,
              "rendered": "Hello",
              "type": "h1",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "122",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "anker mouse",
              " : $",
              50.25,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to Card
              </button>,
              <h1>
                Hello
              </h1>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "anker mouse",
            " : $",
            50.25,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to Card",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to Card",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Hello",
              },
              "ref": null,
              "rendered": "Hello",
              "type": "h1",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "553",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "computer stand",
              " : $",
              20.99,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to Card
              </button>,
              <h1>
                Hello
              </h1>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "computer stand",
            " : $",
            20.99,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to Card",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to Card",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Hello",
              },
              "ref": null,
              "rendered": "Hello",
              "type": "h1",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "689",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "mac adapter",
              " : $",
              10000,
              " USD",
              <button
                onClick={[Function]}
              >
                Add to Card
              </button>,
              <h1>
                Hello
              </h1>,
            ],
            "className": "product-item",
          },
          "ref": null,
          "rendered": Array [
            "mac adapter",
            " : $",
            10000,
            " USD",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add to Card",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add to Card",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Hello",
              },
              "ref": null,
              "rendered": "Hello",
              "type": "h1",
            },
          ],
          "type": "li",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
